### PR TITLE
Refactor classifiers to take numClasses parameters

### PR DESCRIPTION
**Pull request by rcurtin**
This is needed for the cross-validation module.  For each of the mlpack classifiers, I have made sure that the API can match the following signature:

```
void Train(const arma::mat& data, const arma::Row<size_t>& labels, const size_t numClasses);
```

and constructors have been updated appropriately also.  The following classes needed changes:

 - `AdaBoost`
 - `DecisionStump`
 - `LogisticRegression` -- this one is special because logistic regression is two-class only, so I just added an extra overload and corresponding documentation
 - `NaiveBayesClassifier`
 - `Perceptron`

@micyril: take a look at this, let me know if this is what you need or if there are any further changes necessary.

### CODE REVIEW

File: src/mlpack/methods/logistic_regression/logistic_regression.hpp [hunk 1]
```diff
@@ -53,6 +53,30 @@ class LogisticRegression
...
+   * it requires the extra numClasses parameter (which must always be 2).
+   *
+   * @param predictors Input training variables.
+   * @param responses Outputs resulting from input training variables.
+   * @param numClasses Number of classes in the dataset (must be 2).
+   * @param lambda L2-regularization parameter.
+   */
+  LogisticRegression(const MatType& predictors,
+                     const arma::Row<size_t>& responses,
+                     const size_t numClasses,
```
Comment (micyril) [inline src/mlpack/methods/logistic_regression/logistic_regression.hpp]: I guess my comment about lacking the `numClasses` parameter in `LogisticRegression` was a bit missleading. When we disccussed the possibility to add the `numClasses` parameter in CV constructors, I perceived it as we should always require this parameter if it is present in one of `Train` methods. According to this strategy we will require users to pass `numClasses` when they want to apply cross-validation to LogisticRegression. Do we want it? If no, we can leave `LogisticRegression` without changes.

Comment (micyril) [inline src/mlpack/methods/logistic_regression/logistic_regression.hpp]: The same considerations are applied to `AdaBoost`.

File: src/mlpack/methods/naive_bayes/naive_bayes_classifier.hpp [hunk 2]
```diff
@@ -92,11 +92,14 @@ class NaiveBayesClassifier
...
    *
    * @param data The dataset to train on.
+   * @param labels The labels for the dataset.
+   * @param numClasses The numbe of classes in the dataset.
    * @param incremental Whether or not to use the incremental algorithm for
    *      training.
    */
   void Train(const MatType& data,
              const arma::Row<size_t>& labels,
+             const size_t numClasses,
```
Comment (micyril) [inline src/mlpack/methods/naive_bayes/naive_bayes_classifier.hpp]: Is it ok just to add a new parameter to the `Train` method? Shouldn't we deprecate the old method and add a new one?

Comment (rcurtin) [inline src/mlpack/methods/naive_bayes/naive_bayes_classifier.hpp]: That should only be necessary if we plan to release another 2.x.x version with this new signature, but since the next release should be 3.0.0, it is ok to break reverse compatibility at this level.

**Comment by rcurtin** [comment]
Hm, ok, I had thought that we should synchronize the `Train()` methods across all classifiers and then the presence of the `numClasses` parameter could be used to determine whether the method was classification or regression.  Did you have something else in mind?  I thought we can enforce that each classifier must implement

```
ClassifierType(const MatType& data, const arma::Row<size_t>& labels, const size_t numClasses);
void ClassifierType::Train(const MatType& data, const arma::Row<size_t>& labels, const size_t numClasses);
```

but then a question also is what to do with the `DatasetInfo` parameter, which is used by some classifiers that can accept categorical data.  We could have other overloads:

```
ClassifierType(const MatType& data, const DatasetInfo& data, const arma::Row<size_t>& labels, const size_t numClasses);
void ClassifierType::Train(const MatType& data, const DatasetInfo& data, const arma::Row<size_t>& labels, const size_t numClasses);
```

Or, we could force the `DatasetInfo` to be accepted as another later optional parameter to the constructor, i.e., `ClassifierType(const MatType& data, const arma::Row<size_t>& labels, const size_t numClasses, const DatasetInfo& datasetInfo = DatasetInfo())`.

Whichever of those we choose, I think it is easiest if you specify the best interface for your work, and then we can apply it through the codebase.  Since the next release will be 3.0.0 and this allows us to break reverse compatibility, this is the easiest time to make these changes. :)

**Comment by micyril** [comment]
We can support cross-validation for `LogisticRegression` and `AdaBoost` without requirement to pass the unnecessary `numClasses` parameter. The changes in `DecisionStump`, `NaiveBayesClassifier` and `Perceptron` are more appropriate since users always need to pass the `numClasses` parameter into constructors of these classes.
Regarding detection of whether a method is classification or regression - the cross-validation module don't need to know this information as long as it knows which parameters it should require to pass  and which ones it can take as optional.

**Comment by rcurtin** [comment]
I see, thanks for the clarification.  Do you have an opinion of how the APIs for classification and regression methods should be structured?  If we can standardize this, we make it easier to implement new classifiers and regressors.

**Comment by micyril** [comment]
I guess we can use LinearRegression as a de facto standard for regression algorithms, and DecisionTree as a de facto standard for classification algorithms.

**Comment by rcurtin** [comment]
I see, the only issue is that these don't account for categorical variables.  I suggest that we do this:

regression:

```
// I guess usually MatType == arma::Mat<eT> and ResponsesType == arma::Col<eT>
RegressionType(const MatType& data, const ResponsesType& responses, /* all optional arguments past here... */);
void RegressionType::Train(const MatType& data, const ResponsesType& responses);

// Sometimes the input could be categorical for some methods too, so these can be optional:
RegressionType(const MatType& data, const DatasetInfo& info, const ResponsesType& responses, /* all optional arguments past here... */);
void RegressionType::Train(const MatType& data, const DatasetInfo& info, const ResponsesType& responses);
```

classification:

```
ClassificationType(const MatType& data, const arma::Row<size_t>& labels, const size_t numClasses, /* all optional arguments past here... */);
void ClassificationType::Train(const MatType& data, const arma::Row<size_t>& labels, const size_t numClasses);

// optional for classes that support categorical data also
ClassificationType(const MatType& data, const DatasetInfo& info, const arma::Row<size_t>& labels, const size_t numClasses, /* all optional arguments past here... */);
void ClassificationType::Train(const MatType& data, const DatasetInfo& info, const arma::Row<size_t>& labels, const size_t numClasses);
```

In these cases it is not required that the overload is exactly implemented as specified, but the compiler should at least be able to instantiate it with `MatType = arma::mat` and `ResponsesType = arma::vec`.  So it is ok if the signature is implemented a little differently.  It would be nice to eventually require that `MatType` can be an Armadillo object of any type, but there is other support that would be needed first.

@zoq: I wouldn't mind your input here too, this is a fairly fundamental design decision so it would be best to get eyes on this before things start changing.

Whenever we decide on something, I will have to create some documentation detailing exactly what is required of a classification and a regression type, a little like http://mlpack.org/docs/mlpack-2.2.3/doxygen.php?doc=trees.html .

**Comment by micyril** [comment]
For completeness we can also specify an optional parameter for weights.

**Comment by rcurtin** [comment]
Ah, right, I forgot about weights.  I've updated my comment.

File: src/mlpack/methods/logistic_regression/logistic_regression.hpp (hunk 1 above)
Comment (rcurtin) [inline src/mlpack/methods/logistic_regression/logistic_regression.hpp]: Hm, so I spent some time thinking about this point.  I think that in order to preserve compatibility of two-class classifiers with other classifiers, we can simply provide two overloads, one that takes `numClasses` (so that the classifier satisfies the requirements of other classifiers) and one that doesn't, for user convenience.

AdaBoost is a multiclass classifier (or it can be, depending on its weak learner) so I think leaving `numClasses` there is just fine.

Do you think that is ok, or does that cause any problems?

File: src/mlpack/methods/adaboost/adaboost_main.cpp [hunk 3]
```diff
@@ -202,8 +202,10 @@ int main(int argc, char *argv[])
     else if (weakLearner == "perceptron")
       m.WeakLearnerType() = AdaBoostModel::WeakLearnerTypes::PERCEPTRON;
 
+    const size_t numClasses = arma::max(labels) + 1;
```
Comment (zoq) [inline src/mlpack/methods/adaboost/adaboost_main.cpp]: Do we require continuous labels starting with 0? If not I think this doesn't return the correct number of labels.

**Comment by zoq** [comment]
As for me, the proposed interface looks good, as long as we provide the same interface for all methods. I'm a little bit worried about the LogisticRegression method since someone could get the impression that the method works with ``numClasses > 2``, but if can make that clear in the documentation this is just a minor issue. I guess we could also through an exception or something like that in the future, once we decided on the error handling: https://github.com/mlpack/mlpack/issues/891.

File: src/mlpack/methods/adaboost/adaboost_main.cpp (hunk 3 above)
Comment (rcurtin) [inline src/mlpack/methods/adaboost/adaboost_main.cpp]: Yes, that is required, but I am not sure how clear the documentation is.  I'll update it.

Comment (rcurtin) [inline src/mlpack/methods/adaboost/adaboost_main.cpp]: Ack, sorry, I thought this was in a different place than it is.  In this code the call to `NormalizeLabels()` forces the labels to be between 0 and numClasses.  But I can still make the call a little more efficient, because `numClasses == m.Mappings().n_elem`.

**Comment by rcurtin** [comment]
I updated the documentation somewhat to make it clearer that `LogisticRegression` only supports two classes even though the `numClasses` parameter is sometimes available.  If you think that is clear, should we go ahead and merge this?

**Comment by zoq** [comment]
Completely missed the last message, it's clear for me and I think we waited long enough for any more comments, so let's merge this.

**Comment by rcurtin** [comment]
Agreed, so in this case I will go ahead and merge.  If anyone complains we can revisit the issue then :)

File: src/mlpack/methods/logistic_regression/logistic_regression.hpp (hunk 1 above)
Comment (micyril) [inline src/mlpack/methods/logistic_regression/logistic_regression.hpp]: Sorry, I missed this reply. The main disadvantage of keeping `numClasses` in `LogisticRegression` is that users will be required to pass it when they use `LogisticRegression` with the cross-validation and hyper-parameter tuning modules. If we want to keep it for the reason described in the documentation "as necessary for the cross-validation code and hyper-parameter tuner to work", it is not actually the case - cross-validation and hyper-parameter tuning will work without it. So, we need to keep it only if there is some other reason(s). Is there any other code that depends on the assumption of the presence of `numClasses` in classifiers?

Comment (rcurtin) [inline src/mlpack/methods/logistic_regression/logistic_regression.hpp]: I see, so are you saying that we can remove the `numClasses` from `LogisticRegression` and other two-class classifiers entirely, and the functionality of CV or the hyper-parameter tuner will not be affected?  If that is true, I guess we can revert the changes to `LogisticRegression`; would you agree with that?

Comment (micyril) [inline src/mlpack/methods/logistic_regression/logistic_regression.hpp]: Cross-validation and hyper-parameter tuning will work in both cases. The only difference is that in one case it will be required to pass `numClasses`, in another - no. I agree that we can revert the changes for `LogisticRegression`.
